srml/contract/src/account_db.rs

```diff
@@ -178,3 +179,45 @@ impl<'a, T: Trait> AccountDb<T> for OverlayAccountDb<'a, T> {
 		}
 	}
 }
+
+// TODO: Perform clean up.
+// Introduce `storage` module and make `account_db`/`double_map` to be sub-modules of that module
+// and then move this function there (e.g. to the top level). Ideally move the decl_storage part there as well.
+
+/// Transfers all funds from `who` to `inherent` and makes sure all data associated
+/// to `who` is removed.
+pub fn commit_suicide<T: Trait>(who: &T::AccountId, inherent: &T::AccountId) {
+	let residue = balances::Module::<T>::free_balance(who);
+	let inherent_balance = balances::Module::<T>::free_balance(inherent);
+
+	// Increase inherent balance by the suicidee's residue.
+	let new_inherent_balance = inherent_balance.saturating_add(residue);
+	balances::Module::<T>::set_free_balance_creating(inherent, new_inherent_balance);
+
+	// Then nullify the balance of suicidee. This most probably will invoke `OnFreeBalanceZero`
+	// callback.
+	if let balances::UpdateBalanceOutcome::Updated =
+		balances::Module::<T>::set_free_balance(who, T::Balance::sa(0))
```

Comment:
This may actually cause some funds to be completely burned if `who == inherent`, no?

Reply:
Yep, is that a bad thing? What happens in ethereum?